tket/src/Transformations/BasicOptimisation.cpp

```diff
@@ -580,6 +580,132 @@ Transform commute_SQ_gates_through_SWAPS(const op_node_errors_t &node_errors) {
+
+    // Loop through all NPhasedX gates
+    BGL_FORALL_VERTICES(v, circ.dag, DAG) {
+      Op_ptr op = circ.get_Op_ptr_from_Vertex(v);
+      if (op->get_type() == OpType::NPhasedX) {
+        // gather surrounding Rz gates
+        unsigned arity = op->n_qubits();
+        std::vector<Expr> in_rz(arity);
+        std::vector<Expr> out_rz(arity);
+        EdgeVec in_edges = circ.get_in_edges_of_type(v, EdgeType::Quantum);
+        EdgeVec out_edges = circ.get_out_edges_of_type(v, EdgeType::Quantum);
+        TKET_ASSERT(in_edges.size() == arity);
+        TKET_ASSERT(out_edges.size() == arity);
+        for (unsigned i = 0; i < arity; ++i) {
+          Vertex in_v = circ.source(in_edges[i]);
+          Op_ptr in_op = circ.get_Op_ptr_from_Vertex(in_v);
+          Vertex out_v = circ.target(out_edges[i]);
+          Op_ptr out_op = circ.get_Op_ptr_from_Vertex(out_v);
+
+          if (in_op->get_type() == OpType::Rz) {
+            in_rz[i] = -in_op->get_params().at(0);
+          } else {
+            in_rz[i] = 0.;
+          }
+          if (out_op->get_type() == OpType::Rz) {
+            out_rz[i] = out_op->get_params().at(0);
+          } else {
+            out_rz[i] = 0.;
+          }
+        }
+
+        // Find out which Rz angle is most popular.
+        // Note that we only compare expr[i] with expr[j] when j < i. This means
+        // that only the largest i from a set of equivalent exprs will have the
+        // right occurence count, but that is good enough.
+        std::vector<Expr> all_rz = in_rz;
+        all_rz.insert(all_rz.end(), out_rz.begin(), out_rz.end());
+        std::vector<unsigned> occurences_count(2 * arity);
+        for (unsigned i = 0; i < 2 * arity; ++i) {
+          unsigned cnt = 0;
+          for (unsigned j = 0; j < i; ++j) {
+            if (equiv_expr(all_rz[i], all_rz[j], 4)) {
+              ++cnt;
+            }
+          }
+          occurences_count[i] = cnt;
+        }
+        unsigned max_i =
+            std::max_element(occurences_count.begin(), occurences_count.end()) -
+            occurences_count.begin();
+        Expr absorb_rz = all_rz[max_i];
+
+        if (!equiv_0(absorb_rz, 4)) {
+          success = true;
+
+          // Subtract absorb_rz in NPhasedX
+          std::vector<Expr> new_params = op->get_params();
+          TKET_ASSERT(new_params.size() == 2);
+          new_params[1] += absorb_rz;
+          circ.dag[v] = get_op_ptr(OpType::NPhasedX, new_params, arity);
+
+          // Finally, adjust +-absorb_rz in Rz everywhere around
+          for (unsigned i = 0; i < arity; ++i) {
+            Vertex in_v = circ.source(in_edges[i]);
+            Op_ptr in_op = circ.get_Op_ptr_from_Vertex(in_v);
+            Vertex out_v = circ.target(out_edges[i]);
+            Op_ptr out_op = circ.get_Op_ptr_from_Vertex(out_v);
+
+            Expr angle;
+            Edge in_e, out_e;
+            VertexSet bin;
+            if (in_op->get_type() == OpType::Rz) {
```

Comment:
Would it be possible to add the condition that all RZ after executing this pass will be in front or after the NPhasedX?

Reply:
Yes, thank you for explaining that!